Sync with pydoctor 20.12.1 release

mthuurne opened the pull request:
## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10065
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I ran additional checks listed at [Getting Your Patch Accepted](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] I have updated the automated tests.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

mthuurne committed:
Fix docstring of twisted.mail.imap4.MessageSet.last

The `@raises` field takes the exception type as an argument.

This problem wasn't discovered before because pydoctor ignored the
docstring due to incomplete support for properties: the setter
definition would shadow the getter.

mthuurne committed:
Store deprecation notice for `@deprecatedProperty` in pydoctor plugin

Without this, deprecated properties won't have a deprecation notice
in the generated API docs.

mthuurne committed:
Include pydoctor version in "generated by" footer

This syncs a change in pydoctor's templates from commit 0b6e6190.

mthuurne committed:
Move information elements from the page header to body

This syncs a change in pydoctor's templates from commit 4705119a.

mthuurne committed:
Put the documentation category (kind) on non-sticky second line

This syncs a change in pydoctor's templates from commit 86142d61.

mthuurne committed:
Add "Show Private API" button to summary pages

This syncs a change in pydoctor's templates from commit b0ed60b7.

mthuurne committed:
Pass package path to pydoctor as positional argument

The `--add-package` option was deprecated.

(I really wanted to write "parameter" instead of "argument" ;)

mthuurne committed:
Add empty newsfragment

mthuurne committed:
Require pydoctor 20.12.1+ in 'dev_release' extra

mthuurne committed:
Do not run APIBuilderTests with pydoctor before 20.12.1

Twisted's templates require render methods that are not implemented
in pydoctor versions below 20.12.0. But 20.12.0 can fail on systems
that have symlinks in the source path (macOS), so 20.12.1 is required
to make the tests pass in CI on all supported platforms.

mthuurne commented:
The failed "Documentation" and "readthedocs" builds are due to a problem with `pip`'s new resolver, so unrelated to this PR.

The problem on macOS might be an actual problem in pydoctor. I'll investigate it tomorrow.

mthuurne committed:
Update expected source links for package doc pages in test_release

mthuurne commented:
pydoctor 20.12.1 should solve the problem where `Path.relative_to()` raises `ValueError` on macOS.

adiroiban commented on src/twisted/mail/imap4.py:
```diff
@@ -223,8 +223,7 @@ def last(self):
 
-        @raises: L{ValueError} if a largest value has already
-        been set.
+        @raises ValueError: if a largest value has already been set.
```
well... maybe move the docs to setter?

But I think that is best to keep all the property docs includding setter docstring inside the main getter method.

I would say that this is a defect in pydoctor, as pydoctor should observer that we have a property and allow such a docstring..

if we don't have a setter pydoctror can highlight the property as read-only  ... otherwise it will be default read/write.

mthuurne replied:
I think it makes sense to have separate docstrings in the source for getters and setters, but to merge them in the output. I'll move it to the setter.

I agree that the presentation of properties in pydoctor is not ideal, but it is an improvement over ignoring them completely and I didn't want to delay the release even more.

adiroiban commented on src/twisted/python/test/test_release.py:
```diff
@@ -377,7 +377,14 @@ def inner():
+@skipIf(pydoctor is None, "Pydoctor is not present.")
+@skipIf(
+    pydoctor is not None and pydoctor.__version__ < Version("pydoctor", 20, 12, 1),
+    "Pydoctor 20.12.1 or later is required.",
```
do we need  this extra skip?

the setup.cfg asks for a newer pydoctor, so I think that we should be safe.

mthuurne replied:
We need it when running on Python 3.5, where pydoctor is excluded from the `dev_release` extra, but might still be installed anyway.

adiroiban commented on setup.cfg:
```diff
@@ -49,7 +49,7 @@ test =
 dev_release =
     towncrier >= 17.4.0
-    pydoctor == 20.7.0
+    pydoctor >= 20.12.1; python_version >= "3.6"
```
maybe install for 3.7 and up.
The reasoning is that pydoctor is only used by the release manager ... or the automated release process.

We don't really need to support any python version.... and can save some seconds during the automated tests.

mthuurne replied:
The extra isn't just for CI, so I think it should reflect the actual dependencies, not our testing policy. I'm in favor of streamlining CI, since having to wait half an hour for test results is not good. But that should be done by not asking certain tests to be run, rather than pretending they cannot be run.

Also, I don't think we'd get significant speedups from just cutting off a few cases here and there. As is true for optimization in general, we need good profiling tools first, so we can see where the bottlenecks are. Are we not running more in parallel because we don't have the capacity or because the definitions don't allow it? How much time do we spend on actual tests and how much on setting up environments?

One fundamental change we could consider is the matrix approach: we currently test with 4 Python versions (3.5 - 3.8) on 4 operating systems (Windows twice, with different reactor) on every run. It might be sufficient to pick a subset of combinations that does include every Python version and every OS, but not every possible combination. That could halve the number of runs for PRs. There could be a full matrix test run on `trunk` on a daily basis to catch the problems that only occur in specific combinations, but I expect those to be rare.

Anyway, that is a much bigger discussion that should be had on the mailinglist.

adiroiban reviewed (approved):
Nicely done. Thanks. Only minor comments. merge a will :)

mthuurne committed:
Move MessageSet.last docstring to setter

The docstring describes the behavior of the setter, so that is where
it belongs.
